File: leetcode/java/0053. Maximum Subarray.java
package leetcode.java;

// Time complexity: O(N)
// Space complexity: O(1)

class Solution {
    public int maxSubArray(int[] nums) {
        int currSum = 0;
        int maxSum = Integer.MIN_VALUE;
        for (int num : nums) {
            if (currSum < 0) {
                currSum = num;
            } else {
                currSum += num;
            }
            maxSum = currSum > maxSum ? currSum : maxSum;
        }
        return maxSum;
    }
}

File: leetcode/python/0053. Maximum Subarray.py
# Time complexity: O(N)
# Space complexity: O(1)

class Solution(object):
    def maxSubArray(self, nums):
        curr_sum = 0
        maximum = -99999

        for num in nums:
            if curr_sum < 0:
                curr_sum = num
            else:
                curr_sum += num
            maximum = max(maximum, curr_sum)
        return maximum